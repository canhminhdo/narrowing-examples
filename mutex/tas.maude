fmod LOC is
    sorts RsLoc Loc .
    subsort RsLoc < Loc .

    op rs : -> RsLoc [ctor] .
    op cs : -> Loc [ctor] .
endfm

fmod PID is
    sort Pid .
endfm

fmod PROC is
    pr PID .
    pr LOC .
    sorts RsProc Proc .
    subsort RsProc < Proc .

    op (pc[_]:_) : Pid RsLoc -> RsProc [ctor] .
    op (pc[_]:_) : Pid Loc -> Proc [ctor] .
endfm

fmod PROC-SET is
    pr PROC .
    sorts RsProcSet ProcSet .
    subsorts RsProc < RsProcSet < ProcSet .
    subsort Proc < ProcSet .

    op void : -> ProcSet [ctor] .
    op __ : ProcSet ProcSet -> ProcSet [ctor assoc comm id: void] .
    op __ : RsProcSet RsProcSet -> RsProcSet [ctor ditto] .
endfm

mod TAS is
    pr PROC-SET .
    sort Config . 

    op {_|`locked:_} : ProcSet Bool -> Config [ctor] .

    vars RsPSET RsPSET' : RsProcSet .
    vars PSET PSET' : ProcSet .   
    vars P P' : Pid .
    var B : Bool .

    rl [enter] : 
        {(pc[P]: rs) PSET | locked: false}
    =>  {(pc[P]: cs) PSET | locked: true} [narrowing] .

    rl [exit] : 
        {(pc[P]: cs) PSET | locked: B}
    =>  {(pc[P]: rs) PSET | locked: false} [narrowing] .
endm

--- Proving cs reachable
{fold} vu-narrow [1] in TAS : 
    {RsPSET | locked: false} 
=>* {(pc[P]: cs) PSET' | locked: B} .

--- Negatively prove the mutex property
{fold} vu-narrow [1] in TAS : 
    {RsPSET | locked: false} 
=>* {(pc[P]: cs) (pc[P']: cs) PSET' | locked: B} .
