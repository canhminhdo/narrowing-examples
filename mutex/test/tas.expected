==========================================
{fold} vu-narrow [1] in TAS : {RsPSET | locked: false} =>* {(pc[P]: cs) PSET' |
    locked: B} .

state: {pc[@1:Pid]: cs | locked: true}
accumulated substitution:
RsPSET --> pc[@1:Pid]: rs
variant unifier:
PSET' --> void
P --> %1:Pid
B --> true
@1:Pid --> %1:Pid
==========================================
{fold} vu-narrow [1] in TAS : {RsPSET | locked: false} =>* {(pc[P]: cs) (pc[
    P']: cs) PSET' | locked: B} .

No solution.
Bye.
